src/compute/src/sink/persist_sink.rs

```diff
@@ -147,117 +361,365 @@ where
                 .await
                 .expect("could not open persist client");
 
-            if truncate && active_write_worker {
-                truncate_persist_shard(shard_id, &persist_client).await;
-            }
-
             let mut write = persist_client
                 .open_writer::<SourceData, (), Timestamp, Diff>(shard_id)
                 .await
                 .expect("could not open persist shard");
 
             while scheduler.notified().await {
-                let mut input_frontier = Antichain::new();
-                for frontier in frontiers.borrow().clone() {
-                    input_frontier.extend(frontier);
+                if token_weak.upgrade().is_none() {
+                    return;
+                }
+
+                // Extract desired rows as positive contributions to `correction`.
+                desired_input.for_each(|_cap, data| {
+                    data.swap(&mut buffer);
+                    correction.append(&mut buffer);
+                });
+
+                // Extract persist rows as negative contributions to `correction`.
+                persist_input.for_each(|_cap, data| {
+                    data.swap(&mut buffer);
+                    correction.extend(buffer.drain(..).map(|(d, t, r)| (d, t, -r)));
+                });
```

Comment:
Consider changing `correction` to a `Vec<Vec<_>>` instead of a flat `Vec<_>`. The reason for this is that with a flat vector, all data needs to be located in a single allocation, while the nested approach doesn't have this requirement. However, it might make the code further below harder to write, especially consolidating across batches doesn't exist.